feat: Enable monitoring (metrics + alerts + grafana dashboard)

orfeas-k opened the pull request:
Enable metrics in Kubeflow Dashboard (upstream PR https://github.com/kubeflow/kubeflow/pull/7639). More specifically:
* Add `COLLECT_METRICS` envvar to pebble layer
* Add `metrics-endpoint`integrations
* Add prometheus_scrape charm library
* Bump grafana_dashboard charm library
* Bump cosl for compatibility with charm libraries
* Add basic alert rules
* Add grafana dashboard
* Add integration tests

Close #252

#### Resources used for creating the dashboard
* https://www.youtube.com/watch?v=VjFFzGFyVlY
* Dashboard implemented in https://github.com/canonical/kubeflow-volumes-operator/pull/176

#### Testing
1. Deploy
Since this is the Kubeflow dashboard metrics, I 'd recommend deploying whole CKF plus the following in order to check the dashboard observability configuration:
```
alias jd="juju deploy"
alias jrel="juju integrate"
juju refresh kubeflow-dashboard --channel latest/edge/pr-254 --base ubuntu@20.04
jd grafana-agent-k8s --channel latest/stable --trust
jd prometheus-k8s --trust --channel latest/stable
jd grafana-k8s --channel latest/stable --trust
jrel kubeflow-dashboard:metrics-endpoint grafana-agent-k8s
jrel kubeflow-dashboard:grafana-dashboard grafana-agent-k8s
jrel grafana-agent-k8s:send-remote-write prometheus-k8s
jrel grafana-agent-k8s:grafana-dashboards-provider grafana-k8s
jrel grafana-k8s:grafana-source prometheus-k8s:grafana-source  
```
2. Go to prometheus dashboard by navigating to k8s svc :
3. Go to Graph tab and type {juju_charm="kubeflow-dashboard"}. See the metrics populating there. In order to generate web app metrics, you need to hit the web app's endpoint which is doable by navigating to the browser using : of the kubeflow-dashboard k8s service (or hitting it iwht a curl request).
4. Go to Grafana by navigating to its :
5. Get password by
```
juju run grafana/leader get-admin-password
```
7. Login and check the dashboard after hitting the kubeflow dashboard endpoint a few times periodically to imitate real behavior

mvlassis reviewed (dismissed):
I tested on a new deployment on AWS and everything works! The prometheus metrics do register, and the grafana dashboard is updated properly

Suggestion: we may want to also update the `requirements.in` files to include `cosl >= 0.0.56` if that is a hard requirement

NohaIhab commented on src/prometheus_alert_rules/kubeflow_dashboard_requests.rules:
suggestion: add alerts that monitor the node.js process, for example [this one](https://grafana.com/oss/prometheus/exporters/nodejs-exporter/?tab=alerting-rules).

NohaIhab commented on src/grafana_dashboards/kubeflow_dashboard_requests_four_signals.json.tmpl:
suggestion: include dashboard for the nodejs process, maybe similar to [the sample dashboard by Grafana](https://grafana.com/oss/prometheus/exporters/nodejs-exporter/?tab=dashboards)

NohaIhab reviewed (commented):
Thanks @orfeas-k, nicely done. I left a suggestion to monitor the node.js process since I see there are useful metrics we can utilize there:
<details>

```
curl 10.1.97.232:8082/prometheus/metrics
# HELP rest_http_request_duration_seconds REST API response time in seconds
# TYPE rest_http_request_duration_seconds histogram

# HELP rest_http_request_total Total number of HTTP requests for REST API
# TYPE rest_http_request_total counter

# HELP app_info Information about application
# TYPE app_info gauge
app_info{version="0.0.2",app="kubeflow-centraldashboard"} 1

# HELP process_cpu_user_seconds_total Total user CPU time spent in seconds.
# TYPE process_cpu_user_seconds_total counter
process_cpu_user_seconds_total{app="kubeflow-centraldashboard"} 22.815042999999992

# HELP process_cpu_system_seconds_total Total system CPU time spent in seconds.
# TYPE process_cpu_system_seconds_total counter
process_cpu_system_seconds_total{app="kubeflow-centraldashboard"} 4.0750389999999985

# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total{app="kubeflow-centraldashboard"} 26.890082000000007

# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds{app="kubeflow-centraldashboard"} 1741618152

# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes{app="kubeflow-centraldashboard"} 112226304

# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes{app="kubeflow-centraldashboard"} 1745797120

# HELP process_heap_bytes Process heap size in bytes.
# TYPE process_heap_bytes gauge
process_heap_bytes{app="kubeflow-centraldashboard"} 91881472

# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds{app="kubeflow-centraldashboard"} 20

# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds{app="kubeflow-centraldashboard"} 65536

# HELP nodejs_eventloop_lag_seconds Lag of event loop in seconds.
# TYPE nodejs_eventloop_lag_seconds gauge
nodejs_eventloop_lag_seconds{app="kubeflow-centraldashboard"} 0.009567257

# HELP nodejs_eventloop_lag_min_seconds The minimum recorded event loop delay.
# TYPE nodejs_eventloop_lag_min_seconds gauge
nodejs_eventloop_lag_min_seconds{app="kubeflow-centraldashboard"} 0.009109504

# HELP nodejs_eventloop_lag_max_seconds The maximum recorded event loop delay.
# TYPE nodejs_eventloop_lag_max_seconds gauge
nodejs_eventloop_lag_max_seconds{app="kubeflow-centraldashboard"} 0.024920063

# HELP nodejs_eventloop_lag_mean_seconds The mean of the recorded event loop delays.
# TYPE nodejs_eventloop_lag_mean_seconds gauge
nodejs_eventloop_lag_mean_seconds{app="kubeflow-centraldashboard"} 0.010273717724531378

# HELP nodejs_eventloop_lag_stddev_seconds The standard deviation of the recorded event loop delays.
# TYPE nodejs_eventloop_lag_stddev_seconds gauge
nodejs_eventloop_lag_stddev_seconds{app="kubeflow-centraldashboard"} 0.0007219516186417628

# HELP nodejs_eventloop_lag_p50_seconds The 50th percentile of the recorded event loop delays.
# TYPE nodejs_eventloop_lag_p50_seconds gauge
nodejs_eventloop_lag_p50_seconds{app="kubeflow-centraldashboard"} 0.010125311

# HELP nodejs_eventloop_lag_p90_seconds The 90th percentile of the recorded event loop delays.
# TYPE nodejs_eventloop_lag_p90_seconds gauge
nodejs_eventloop_lag_p90_seconds{app="kubeflow-centraldashboard"} 0.010420223

# HELP nodejs_eventloop_lag_p99_seconds The 99th percentile of the recorded event loop delays.
# TYPE nodejs_eventloop_lag_p99_seconds gauge
nodejs_eventloop_lag_p99_seconds{app="kubeflow-centraldashboard"} 0.012976127

# HELP nodejs_active_handles Number of active libuv handles grouped by handle type. Every handle type is C++ class name.
# TYPE nodejs_active_handles gauge
nodejs_active_handles{type="Socket",app="kubeflow-centraldashboard"} 3
nodejs_active_handles{type="Server",app="kubeflow-centraldashboard"} 1

# HELP nodejs_active_handles_total Total number of active handles.
# TYPE nodejs_active_handles_total gauge
nodejs_active_handles_total{app="kubeflow-centraldashboard"} 4

# HELP nodejs_active_requests Number of active libuv requests grouped by request type. Every request type is C++ class name.
# TYPE nodejs_active_requests gauge

# HELP nodejs_active_requests_total Total number of active requests.
# TYPE nodejs_active_requests_total gauge
nodejs_active_requests_total{app="kubeflow-centraldashboard"} 0

# HELP nodejs_heap_size_total_bytes Process heap size from Node.js in bytes.
# TYPE nodejs_heap_size_total_bytes gauge
nodejs_heap_size_total_bytes{app="kubeflow-centraldashboard"} 35938304

# HELP nodejs_heap_size_used_bytes Process heap size used from Node.js in bytes.
# TYPE nodejs_heap_size_used_bytes gauge
nodejs_heap_size_used_bytes{app="kubeflow-centraldashboard"} 33871576

# HELP nodejs_external_memory_bytes Node.js external memory size in bytes.
# TYPE nodejs_external_memory_bytes gauge
nodejs_external_memory_bytes{app="kubeflow-centraldashboard"} 1855117

# HELP nodejs_heap_space_size_total_bytes Process heap space size total from Node.js in bytes.
# TYPE nodejs_heap_space_size_total_bytes gauge
nodejs_heap_space_size_total_bytes{space="read_only",app="kubeflow-centraldashboard"} 262144
nodejs_heap_space_size_total_bytes{space="new",app="kubeflow-centraldashboard"} 1048576
nodejs_heap_space_size_total_bytes{space="old",app="kubeflow-centraldashboard"} 23150592
nodejs_heap_space_size_total_bytes{space="code",app="kubeflow-centraldashboard"} 430080
nodejs_heap_space_size_total_bytes{space="map",app="kubeflow-centraldashboard"} 1839104
nodejs_heap_space_size_total_bytes{space="large_object",app="kubeflow-centraldashboard"} 9158656
nodejs_heap_space_size_total_bytes{space="code_large_object",app="kubeflow-centraldashboard"} 49152
nodejs_heap_space_size_total_bytes{space="new_large_object",app="kubeflow-centraldashboard"} 0

# HELP nodejs_heap_space_size_used_bytes Process heap space size used from Node.js in bytes.
# TYPE nodejs_heap_space_size_used_bytes gauge
nodejs_heap_space_size_used_bytes{space="read_only",app="kubeflow-centraldashboard"} 32808
nodejs_heap_space_size_used_bytes{space="new",app="kubeflow-centraldashboard"} 374992
nodejs_heap_space_size_used_bytes{space="old",app="kubeflow-centraldashboard"} 22789592
nodejs_heap_space_size_used_bytes{space="code",app="kubeflow-centraldashboard"} 295104
nodejs_heap_space_size_used_bytes{space="map",app="kubeflow-centraldashboard"} 1260000
nodejs_heap_space_size_used_bytes{space="large_object",app="kubeflow-centraldashboard"} 9119920
nodejs_heap_space_size_used_bytes{space="code_large_object",app="kubeflow-centraldashboard"} 2784
nodejs_heap_space_size_used_bytes{space="new_large_object",app="kubeflow-centraldashboard"} 0

# HELP nodejs_heap_space_size_available_bytes Process heap space size available from Node.js in bytes.
# TYPE nodejs_heap_space_size_available_bytes gauge
nodejs_heap_space_size_available_bytes{space="read_only",app="kubeflow-centraldashboard"} 0
nodejs_heap_space_size_available_bytes{space="new",app="kubeflow-centraldashboard"} 672464
nodejs_heap_space_size_available_bytes{space="old",app="kubeflow-centraldashboard"} 155936
nodejs_heap_space_size_available_bytes{space="code",app="kubeflow-centraldashboard"} 35360
nodejs_heap_space_size_available_bytes{space="map",app="kubeflow-centraldashboard"} 572184
nodejs_heap_space_size_available_bytes{space="large_object",app="kubeflow-centraldashboard"} 0
nodejs_heap_space_size_available_bytes{space="code_large_object",app="kubeflow-centraldashboard"} 0
nodejs_heap_space_size_available_bytes{space="new_large_object",app="kubeflow-centraldashboard"} 1047456

# HELP nodejs_version_info Node.js version info.
# TYPE nodejs_version_info gauge
nodejs_version_info{version="v12.22.9",major="12",minor="22",patch="9",app="kubeflow-centraldashboard"} 1

# HELP nodejs_gc_duration_seconds Garbage collection duration by kind, one of major, minor, incremental or weakcb.
# TYPE nodejs_gc_duration_seconds histogram
nodejs_gc_duration_seconds_bucket{le="0.001",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="0.01",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="0.1",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="1",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="2",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="5",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="+Inf",app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_sum{app="kubeflow-centraldashboard",kind="incremental"} 0.00026420299999999996
nodejs_gc_duration_seconds_count{app="kubeflow-centraldashboard",kind="incremental"} 2
nodejs_gc_duration_seconds_bucket{le="0.001",app="kubeflow-centraldashboard",kind="major"} 0
nodejs_gc_duration_seconds_bucket{le="0.01",app="kubeflow-centraldashboard",kind="major"} 0
nodejs_gc_duration_seconds_bucket{le="0.1",app="kubeflow-centraldashboard",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="1",app="kubeflow-centraldashboard",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="2",app="kubeflow-centraldashboard",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="5",app="kubeflow-centraldashboard",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="+Inf",app="kubeflow-centraldashboard",kind="major"} 2
nodejs_gc_duration_seconds_sum{app="kubeflow-centraldashboard",kind="major"} 0.025477856
nodejs_gc_duration_seconds_count{app="kubeflow-centraldashboard",kind="major"} 2
nodejs_gc_duration_seconds_bucket{le="0.001",app="kubeflow-centraldashboard",kind="minor"} 20
nodejs_gc_duration_seconds_bucket{le="0.01",app="kubeflow-centraldashboard",kind="minor"} 32
nodejs_gc_duration_seconds_bucket{le="0.1",app="kubeflow-centraldashboard",kind="minor"} 32
nodejs_gc_duration_seconds_bucket{le="1",app="kubeflow-centraldashboard",kind="minor"} 32
nodejs_gc_duration_seconds_bucket{le="2",app="kubeflow-centraldashboard",kind="minor"} 32
nodejs_gc_duration_seconds_bucket{le="5",app="kubeflow-centraldashboard",kind="minor"} 32
nodejs_gc_duration_seconds_bucket{le="+Inf",app="kubeflow-centraldashboard",kind="minor"} 32
nodejs_gc_duration_seconds_sum{app="kubeflow-centraldashboard",kind="minor"} 0.038189219999999996
nodejs_gc_duration_seconds_count{app="kubeflow-centraldashboard",kind="minor"} 32
```

</details>

orfeas-k commented:
@NohaIhab thank you for pointing to those resources. I updated the PR with:
* a new dashboard for nodejs metrics
* updated fields in requests dashboard for unified appearance
* updated fields in requests tracking "requests under 500ms" after our offline discussion
* updated fields in requests regarding memory and cpu usage for unified approach between the two dashboard
* Added one alert rule (inspired from the linked one) for checking that the nodejs process is running

NohaIhab reviewed (approved):
thank you @orfeas-k 
I tested by viewing the alerts and the dashboard in a CKF deployment, dashboard looks great:
![Screenshot from 2025-03-12 14-23-36](https://github.com/user-attachments/assets/b6cab653-c58c-4759-b8ea-56e4db74d524)
